### PR TITLE
fix(bundler): fix recursive resolved dependencies

### LINKED ISSUE #15949: Error trying to run bundle lock --update undefined
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.78.1

### Please select which platform you are using if self-hosting.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitHub.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

* Renovate is trying to upgrade the `rails` gem from `6.0.5` to `6.1.6` using `rangeStrategy=update-lockfile`.
* We have another gem, `roadie-rails`, defined in the `Gemfile` with no version specified, but currently on [`2.1.1`](https://rubygems.org/gems/roadie-rails/versions/2.1.1) in the `Gemfile.lock`. This version defines `rails < 6.1`.
* The upgrade fails because of `roadie-rails`, as expected, but on the pull request it adds a comment with "Artifact update problem" that specifies:

File name: Gemfile.lock

```
Could not find gem 'undefined'.
```

I would expect `roadie-rails` to be mentioned somewhere, as seen when I try to run the same myself:

```
❯ bundler lock --update rails
Fetching gem metadata from https://rubygems.org/...........
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "railties":
  In Gemfile:
    rails (= 6.1.6) was resolved to 6.1.6, which depends on
      railties (= 6.1.6)

    roadie-rails was resolved to 2.1.1, which depends on
      railties (>= 5.1, < 6.1)
```

The logs seem to show that renovate detects a problem updating rails, even finds that `roadie-rails` is the problem but then tries to do a `bundler lock --update rails undefined undefined`.

Here's the minimal reproduction: https://github.com/mrsimo/renovate-bug-undefined

To double check that it wasn't a problem with two private gems we have I removed them, but the problem still occurs in that repository as you might see [from the two PR's my test run created](https://github.com/mrsimo/renovate-bug-undefined/pulls).

### Relevant debug logs

I think this is the relevant bits. I redacted the `updatedDeps` from the `Found new resolve matches - reattempting recursively` because it didn't seem to have anything useful. That line does say `"resolveMatches":["rails","roadie-rails"]`.

<details><summary>Logs</summary>

```json
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","command":"bundler lock --update rails","msg":"Executing command","time":"2022-06-08T15:30:51.817Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","err":{"code":6,"killed":false,"signal":null,"cmd":"bundler lock --update rails","stdout":"Fetching https://**redacted**@github.com/harvesthq/harvest-revision.git\nFetching https://**redacted**@github.com/harvesthq/harvest-lb-check.git\nFetching gem metadata from https://rubygems.org/..........\nResolving dependencies....\n","stderr":"Bundler could not find compatible versions for gem \"railties\":\n  In Gemfile:\n    rails (= 6.1.6) was resolved to 6.1.6, which depends on\n      railties (= 6.1.6)\n\n    roadie-rails was resolved to 2.1.1, which depends on\n      railties (>= 5.1, < 6.1)\n","message":"Command failed: bundler lock --update rails\nBundler could not find compatible versions for gem \"railties\":\n  In Gemfile:\n    rails (= 6.1.6) was resolved to 6.1.6, which depends on\n      railties (= 6.1.6)\n\n    roadie-rails was resolved to 2.1.1, which depends on\n      railties (>= 5.1, < 6.1)\n","stack":"Error: Command failed: bundler lock --update rails\nBundler could not find compatible versions for gem \"railties\":\n  In Gemfile:\n    rails (= 6.1.6) was resolved to 6.1.6, which depends on\n      railties (= 6.1.6)\n\n    roadie-rails was resolved to 2.1.1, which depends on\n      railties (>= 5.1, < 6.1)\n\n    at ChildProcess.exithandler (node:child_process:398:12)\n    at ChildProcess.emit (node:events:527:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"},"msg":"rawExec err","time":"2022-06-08T15:30:56.204Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","err":{"code":6,"killed":false,"signal":null,"cmd":"bundler lock --update rails","stdout":"Fetching https://**redacted**@github.com/harvesthq/harvest-revision.git\nFetching https://**redacted**@github.com/harvesthq/harvest-lb-check.git\nFetching gem metadata from https://rubygems.org/..........\nResolving dependencies....\n","stderr":"Bundler could not find compatible versions for gem \"railties\":\n  In Gemfile:\n    rails (= 6.1.6) was resolved to 6.1.6, which depends on\n      railties (= 6.1.6)\n\n    roadie-rails was resolved to 2.1.1, which depends on\n      railties (>= 5.1, < 6.1)\n","message":"Command failed: bundler lock --update rails\nBundler could not find compatible versions for gem \"railties\":\n  In Gemfile:\n    rails (= 6.1.6) was resolved to 6.1.6, which depends on\n      railties (= 6.1.6)\n\n    roadie-rails was resolved to 2.1.1, which depends on\n      railties (>= 5.1, < 6.1)\n","stack":"Error: Command failed: bundler lock --update rails\nBundler could not find compatible versions for gem \"railties\":\n  In Gemfile:\n    rails (= 6.1.6) was resolved to 6.1.6, which depends on\n      railties (= 6.1.6)\n\n    roadie-rails was resolved to 2.1.1, which depends on\n      railties (>= 5.1, < 6.1)\n\n    at ChildProcess.exithandler (node:child_process:398:12)\n    at ChildProcess.emit (node:events:527:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"},"msg":"Bundler has a resolve error","time":"2022-06-08T15:30:56.205Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","resolveMatches":["rails","roadie-rails"],"updatedDeps":["**redacted**"],"msg":"Found new resolve matches - reattempting recursively","time":"2022-06-08T15:30:56.205Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","msg":"bundler.updateArtifacts(Gemfile)","time":"2022-06-08T15:30:56.209Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","msg":"Using bundler version specified in lockfile","time":"2022-06-08T15:30:56.210Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","msg":"Using ruby version specified in .ruby-version","time":"2022-06-08T15:30:56.211Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","command":"bundler lock --update rails undefined undefined","msg":"Executing command","time":"2022-06-08T15:30:56.212Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":20,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","err":{"code":7,"killed":false,"signal":null,"cmd":"bundler lock --update rails undefined undefined","stdout":"","stderr":"Could not find gem 'undefined'.\n","message":"Command failed: bundler lock --update rails undefined undefined\nCould not find gem 'undefined'.\n","stack":"Error: Command failed: bundler lock --update rails undefined undefined\nCould not find gem 'undefined'.\n\n    at ChildProcess.exithandler (node:child_process:398:12)\n    at ChildProcess.emit (node:events:527:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"},"msg":"rawExec err","time":"2022-06-08T15:30:56.549Z","v":0}
{"name":"renovate","hostname":"renovate-27578370-pc7dt","pid":8,"level":30,"logContext":"8Hq1HOVjZdpBW_sRpYtam","repository":"harvesthq/paperboy","branch":"renovate-rails-6.x","err":{"code":7,"killed":false,"signal":null,"cmd":"bundler lock --update rails undefined undefined","stdout":"","stderr":"Could not find gem 'undefined'.\n","message":"Command failed: bundler lock --update rails undefined undefined\nCould not find gem 'undefined'.\n","stack":"Error: Command failed: bundler lock --update rails undefined undefined\nCould not find gem 'undefined'.\n\n    at ChildProcess.exithandler (node:child_process:398:12)\n    at ChildProcess.emit (node:events:527:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"},"msg":"Gemfile.lock update failed due to an unknown reason","time":"2022-06-08T15:30:56.550Z","v":0}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Add more tests
- Fix types and code

## Context

- closes #15949

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Example: https://github.com/renovate-testing/test-15949-renovate-bundler-recursive/pull/4

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:470]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce('Updated Gemfile.lock');
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:517]:
```suggestion
      fs.writeLocalFile.mockResolvedValueOnce();
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:538]:
```suggestion
      fs.writeLocalFile.mockResolvedValueOnce();
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:582]:
```suggestion
```

File: lib/modules/manager/bundler/artifacts.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.ts:84]: better use `is.nonEmptyOrWhitespaceString`

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:192]:
```suggestion
```
can probably removed too

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:229]:
```suggestion
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:271]:
```suggestion
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:313]:
```suggestion
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:362]:
```suggestion
```

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.spec.ts:630]:
```suggestion
```

**Review by viceice — COMMENTED** [review]
let's remove all dummy resolve, the default mock should do it for tests

File: lib/modules/manager/bundler/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/bundler/artifacts.ts:53]: Any chance this could enter an infinite loop? Maybe check if `pkg` is already in `result` and break out of the loop if so?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/bundler/artifacts.ts
Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.ts:53]: Checking for infinite loop is always a good idea 👍

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/bundler/artifacts.ts
Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.ts:53]: I'm not sure about infinite recursion criteria, so instead I've limited nested calls by 10 at maximum

Comment (rarkins) [reply lib/modules/manager/bundler/artifacts.ts:53]: So it's impossible that this `while` loop won't go infinitely? What about checking for a repeat of same `pkg`?

Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.ts:53]: From iteration to iteration, we extend `updatedDeps` list with at least one new package name which is obtained from the ouside source. Seems like depth limit is the best we can do here.

File: lib/modules/manager/bundler/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/bundler/artifacts.ts:55]:
```suggestion
    if (result.includes(pkg)) {
        logger.debug('pkg already found in result');
        break;
    }
    result.push(pkg);
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/bundler/artifacts.ts
Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.ts:55]: Added uniqueness instead. I'm not sure early break would protect from infinite loop.

File: lib/modules/manager/bundler/artifacts.ts
Comment (viceice) [inline lib/modules/manager/bundler/artifacts.ts:52]: i think this will cause issues if you pass exactly the same string twice to this function, the second time it won't find anything.

can we split the string by newlines instead and iterate over?

the big issue is: do not reuse cached regex, as it stores state

File: lib/modules/manager/bundler/artifacts.ts
Comment (viceice) [reply lib/modules/manager/bundler/artifacts.ts:53]: i don't see this resolved, see my other comment

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.spec.ts:54]: Let's wait for #16414 to be merged

File: lib/modules/manager/bundler/artifacts.ts
Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.ts]: Oh sorry, confused this with another potential infinite loop case in this PR

**Review by viceice — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

**Comment by zharinov** [comment]
😬

File: lib/modules/manager/bundler/artifacts.spec.ts
Comment (rarkins) [reply lib/modules/manager/bundler/artifacts.spec.ts:54]: Marking this as draft until the other is merged

Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.spec.ts]: > move to exec utils to reuse in other tests?

Is this offer still valid? I've just added `exec-error.ts` from #16414, so we don't have to wait

Comment (zharinov) [reply lib/modules/manager/bundler/artifacts.spec.ts]: Ah, never mind
image

**Review by viceice — APPROVED** [review]

**Pull request merged**
